### ISSUE TITLE
8-bit / 4-bit in `diffusers`?

### ISSUE THREAD

**Issue by sayakpaul** [huggingface/diffusers#6500]
## TL;DR

This issue is about gathering interest for supporting 8-bit and 4-bit precision operations (via `bitsandbytes`) in `diffusers` like `transformers`. It does so by showing some preliminary code and discussing the gotchas we should be aware of. It's recommended to read the entirety of the description first, and then comment. If you feel it's worth supporting this feature in `diffusers`, please leave a 👍. 

## Why support it?

Same reasons why we support it in `transformers`:

* Improve accessibility of large models on very low-resource devices.
* Trainability on resource-constrained environments. 

<details>
<summary><h2>Initial investigation</h2></summary>

With the help of @younesbelkada, I came up with a quick PoC to allow performing inference in 8bit and 4bit. Here's the script:

```python
from diffusers import DiffusionPipeline
import argparse
import torch
import time
import bitsandbytes as bnb
import json

SHORT_NAME_MAPPER = {
    "runwayml/stable-diffusion-v1-5": "sd-v1.5",
    "stabilityai/stable-diffusion-xl-base-1.0": "sdxl"
}


def load_pipeline(args):
    pipeline = DiffusionPipeline.from_pretrained(args.ckpt_id, torch_dtype=torch.float16).to("cuda")

    def replace_regular_linears(module, mode="8bit"):
        for name, child in module.named_children():
            if isinstance(child, torch.nn.Linear):
                in_features = child.in_features
                out_features = child.out_features
                device = child.weight.data.device

                # Create and configure the Linear layer
                has_bias = True if child.bias is not None else False
                if mode == "8bit":
                    new_layer = bnb.nn.Linear8bitLt(in_features, out_features, bias=has_bias, has_fp16_weights=False)
                else:
                    # TODO: Make that configurable
                    # fp16 for compute dtype leads to faster inference
                    # and one should almost always use nf4 as a rule of thumb
                    bnb_4bit_compute_dtype = torch.float16
                    quant_type = "nf4"

                    new_layer = bnb.nn.Linear4bit(
                        in_features,
                        out_features,
                        bias=has_bias,
                        compute_dtype=bnb_4bit_compute_dtype,
                        quant_type=quant_type,
                    )

                new_layer.load_state_dict(child.state_dict())
                new_layer = new_layer.to(device)

                # Set the attribute
                setattr(module, name, new_layer)
            else:
                # Recursively apply to child modules
                replace_regular_linears(child, mode=mode)

    if args.mode is not None:
        replace_regular_linears(pipeline.unet, args.mode)

    pipeline.set_progress_bar_config(disable=True)
    return pipeline


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--ckpt_id", default="runwayml/stable-diffusion-v1-5", type=str, choices=list(SHORT_NAME_MAPPER.keys()))
    parser.add_argument("--mode", default=None, type=str, choices=["8bit", "4bit"])
    parser.add_argument("--prompt", default="a golden vase with different flowers", type=str)
    parser.add_argument("--save_unet", action="store_true")
    args = parser.parse_args()

    torch.cuda.reset_peak_memory_stats()
    pipeline = load_pipeline(args)
    _ = pipeline(args.prompt, generator=torch.manual_seed(2024))

    start = time.time()
    output = pipeline(args.prompt, generator=torch.manual_seed(2024))
    end = time.time()
    mem_bytes = torch.cuda.max_memory_allocated()

    image = output.images[0]
    image_name = f"{SHORT_NAME_MAPPER[args.ckpt_id]}" + "_".join(args.prompt.split(" ")) 
    if args.mode is not None:
        image_name += f"_{args.mode}"
    image.save(f"{image_name}.png")

    print(f"Memory: {mem_bytes/(10**6):.3f} MB")
    print(f"Execution time: {(end - start):.3f} sec")

    info = dict(memory=f"{mem_bytes/(10**6):.3f}", time=f"{(end - start):.3f}")
    info_file = f"{SHORT_NAME_MAPPER[args.ckpt_id]}_info.json" if args.mode is None else f"{SHORT_NAME_MAPPER[args.ckpt_id]}_info_{args.mode}.json"
    with open(info_file, "w") as f:
        json.dump(info, f)

    if args.save_unet:
        save_path = f"{SHORT_NAME_MAPPER[args.ckpt_id]}" + "_unet"
        if args.mode is not None:
            save_path += f"_{args.mode}"
        pipeline.unet.save_pretrained(save_path)
```

</details>

<details>
<summary><h2>Results</h2></summary>

### Images

<table>
    <tr>
        <th style="text-align:center;">model</th>
        <th style="text-align:center;">fp16</th>
        <th style="text-align:center;">8bit</th>
        <th style="text-align:center;">4bit</th>
    </tr>
    <tr>
        <td style="text-align:center;">SD (v1.5)</td>
        <td style="text-align:center;"><img src="https://github.com/huggingface/diffusers/assets/22957388/7a3817bb-225c-4bc3-8312-a56166232adc" width=200/></td>
        <td style="text-align:center;"><img src="https://github.com/huggingface/diffusers/assets/22957388/e5117851-6880-48c8-bde1-46b2bd1cbc98" width=200/></td>
        <td style="text-align:center;"><img src="https://github.com/huggingface/diffusers/assets/22957388/68a70ca2-742b-4809-bb63-884a69d45d6c" width=200/></td>
    </tr>
    <tr>
        <td style="text-align:center;">SDXL</td>
        <td style="text-align:center;"><img src="https://github.com/huggingface/diffusers/assets/22957388/0b1201b2-bc6f-47a6-9f7c-d3afc9c33d02" width=200/></td>
        <td style="text-align:center;"><img src="https://github.com/huggingface/diffusers/assets/22957388/cacb1972-de08-4c49-b2e6-3c372bfd089b" width=200/></td>
        <td style="text-align:center;"><img src="https://github.com/huggingface/diffusers/assets/22957388/fde42321-8943-4671-9ea4-f27285008572" width=200/></td>
    </tr>
</table>

_(Click on the images to view the originals)_

### Memory and speed

**SD (v1.5)**

<table>
    <tr>
        <th style="text-align:center;">variant</th>
        <th style="text-align:center;">memory (MB)</th>
        <th style="text-align:center;">inference (secs)</th>
    </tr>
    <tr>
        <td style="text-align:center;">fp16</td>
        <td style="text-align:center;">3427.855</td>
        <td style="text-align:center;">1.832</td>
    </tr>
    <tr>
        <td style="text-align:center;">8bit</td>
        <td style="text-align:center;">3149.633</td>
        <td style="text-align:center;">5.300</td>
    </tr>
    <tr>
        <td style="text-align:center;">4bit</td>
        <td style="text-align:center;">3012.030</td>
        <td style="text-align:center;">3.420</td>
    </tr>
</table>

**SDXL**

<table>
    <tr>
        <th style="text-align:center;">variant</th>
        <th style="text-align:center;">memory (MB)</th>
        <th style="text-align:center;">inference (secs)</th>
    </tr>
    <tr>
        <td style="text-align:center;">fp16</td>
        <td style="text-align:center;">11237.171</td>
        <td style="text-align:center;">6.614</td>
    </tr>
    <tr>
        <td style="text-align:center;">8bit</td>
        <td style="text-align:center;">9003.258</td>
        <td style="text-align:center;">19.120</td>
    </tr>
    <tr>
        <td style="text-align:center;">4bit</td>
        <td style="text-align:center;">7868.401</td>
        <td style="text-align:center;">11.379</td>
    </tr>
</table>

Note that we didn't apply quantization to the VAE. Doing so should lead to further savings in memory. The memory savings come at the cost of increased latency, which is well-known. 

</details>

<details>
<summary><h2>Offloading + quantization</h2></summary>

The limitation of using quantization during offloading is that if you do CPU / disk offloading, you can’t keep the offloaded linear layers quantized. So, there might be a little conversion overhead when loading offloaded layers back into memory when using quantization. 

I think offloading support should be seamless because many people can only run a pipeline with offloading enabled. More so because a pipeline is composed of multiple models, unlike `transformers`. 

Other way to tackle this would be to run the models underlying a pipeline individually with quantitation and manually bind their outputs sequentially as typically done inside of a pipeline. But this isn't a seamless dev experience. So, I vote against it. 
</details>

<details>
<summary><h2>Design</h2></summary>

We should first add support for the models and then for pipelines. It will closely follow the details from `transformers`: https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py. 

This will be most beneficial for models purely/mostly composed of `nn.Linear` layers as `bitsandbytes` support is limited to `nn.Linear` layers only. 

I did a study to see the proportion of linear and conv layers that SD v1.5 and SDXL UNets have. This is what I found out:

<table>
    <tr>
        <th style="text-align:center;">model</th>
        <th style="text-align:center;">nn.Conv2d</th>
        <th style="text-align:center;">nn.Linear</th>
    </tr>
    <tr>
        <td style="text-align:center;">SD (v1.5)</td>
        <td style="text-align:center;">0.2</td>
        <td style="text-align:center;">0.38</td>
    </tr>
    <tr>
        <td style="text-align:center;">SDXL</td>
        <td style="text-align:center;">0.04</td>
        <td style="text-align:center;">0.57</td>
    </tr>
</table>

Colab Notebook for the above study is [here](https://colab.research.google.com/gist/sayakpaul/a8c913409066bfaf5849279bf1e3f00b/scratchpad.ipynb). 

</details> 

## What's your point?

Does it make sense to support 8bit/4bit inference and training in `diffusers`? It's gonna be a long fight but I think definitely worth fighting for. 

Cc: @younesbelkada @patrickvonplaten @DN6

**Comment by gzguevara** [comment]
Not long ago I listen to this talk about [Q Diffusion](https://www.youtube.com/watch?v=virARwF_pt4&t=1669s). There a ton of interesting details in quantisation of diffusion models. @sayakpaul, as you said "It's gonna be a long fight" haha But I would be amazing!!

**Comment by patrickvonplaten** [comment]
Thanks a lot for the write-up! 
I'd be very interested in a comparisons of 8bit/4bit vs. fp16 with model offloading. If fp16 with model offloading is both faster and requires less memory than, I'm not super sure whether this should be a high prio issue

**Comment by sayakpaul** [comment]
Indeed!

Then we also have titled and sliced decoding with the VAE. I will run a few numbers and get back here.

**Comment by sayakpaul** [comment]
Okay, so I tested the effectiveness of model CPU offload with VAE slicing turned on and off. 

<details>
<summary>My code</summary>

```python
from diffusers import DiffusionPipeline
import argparse
import torch
import time
import bitsandbytes as bnb
import json

SHORT_NAME_MAPPER = {
    "runwayml/stable-diffusion-v1-5": "sd-v1.5",
    "stabilityai/stable-diffusion-xl-base-1.0": "sdxl"
}


def load_pipeline(args):
    if args.enable_model_cpu_offload:
        pipeline = DiffusionPipeline.from_pretrained(args.ckpt_id, torch_dtype=torch.float16)
        if args.enable_vae_slicing:
            pipeline.enable_vae_slicing()
        pipeline.enable_model_cpu_offload()
    else:
        pipeline = DiffusionPipeline.from_pretrained(args.ckpt_id, torch_dtype=torch.float16).to("cuda")
        if args.enable_vae_slicing:
            pipeline.enable_vae_slicing()

    pipeline.set_progress_bar_config(disable=True)
    return pipeline


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--ckpt_id", default="runwayml/stable-diffusion-v1-5", type=str, choices=list(SHORT_NAME_MAPPER.keys()))
    parser.add_argument("--prompt", default="a golden vase with different flowers", type=str)
    parser.add_argument("--enable_model_cpu_offload", action="store_true")
    parser.add_argument("--enable_vae_slicing", action="store_true")
    args = parser.parse_args()

    torch.cuda.reset_peak_memory_stats()
    pipeline = load_pipeline(args)
    _ = pipeline(args.prompt, generator=torch.manual_seed(2024))

    start = time.time()
    output = pipeline(args.prompt, generator=torch.manual_seed(2024))
    end = time.time()
    mem_bytes = torch.cuda.max_memory_allocated()

    image = output.images[0]
    image_name = f"{SHORT_NAME_MAPPER[args.ckpt_id]}" + "_".join(args.prompt.split(" ")) 
    image_name += f"-mco@{args.enable_model_cpu_offload}" + f"-vs@{args.enable_vae_slicing}"
    image.save(f"{image_name}.png")

    print(f"Memory: {mem_bytes/(10**6):.3f} MB")
    print(f"Execution time: {(end - start):.3f} sec")

    info = dict(memory=f"{mem_bytes/(10**6):.3f}", time=f"{(end - start):.3f}")
    info_file = f"{SHORT_NAME_MAPPER[args.ckpt_id]}_info"
    info_file += f"-mco@{args.enable_model_cpu_offload}" + f"-vs@{args.enable_vae_slicing}"
    with open(f"{info_file}.json", "w") as f:
        json.dump(info, f)
```
</details>

<details>
<summary>Bash script to run a bulk of experiments:</summary>
```bash
#!/bin/bash

# Define the ckpt_id values
ckpt_ids=("runwayml/stable-diffusion-v1-5" "stabilityai/stable-diffusion-xl-base-1.0")

# Enable or disable model cpu offload
cpu_offloads=("--enable_model_cpu_offload" "")

# Enable or disable vae slicing
vae_slicings=("--enable_vae_slicing" "")

# Iterate over each combination of options
for ckpt_id in "${ckpt_ids[@]}"; do
    for cpu_offload in "${cpu_offloads[@]}"; do
        for vae_slicing in "${vae_slicings[@]}"; do
            # Construct the command
            cmd="python run_with_offloading.py --ckpt_id=$ckpt_id"
            
            if [ -n "$cpu_offload" ]; then
                cmd="$cmd $cpu_offload"
            fi

            if [ -n "$vae_slicing" ]; then
                cmd="$cmd $vae_slicing"
            fi

            # Run the command
            echo "Running: $cmd"
            $cmd
        done
    done
done
```
</details>

## Findings

### SD 

<table>
    <tr>
        <th style="text-align:center;">setting</th>
        <th style="text-align:center;">memory (MB)</th>
        <th style="text-align:center;">inference (secs)</th>
    </tr>
    <tr>
        <td style="text-align:center;">no offloading + no slicing</td>
        <td style="text-align:center;">3427.855</td>
        <td style="text-align:center;">1.322</td>
    </tr>
    <tr>
        <td style="text-align:center;">no offloading + slicing</td>
        <td style="text-align:center;">3418.828</td>
        <td style="text-align:center;">1.326</td>
    </tr>
    <tr>
        <td style="text-align:center;">offloading + no slicing</td>
        <td style="text-align:center;">2526.685</td>
        <td style="text-align:center;">4.078</td>
    </tr>
    <tr>
        <td style="text-align:center;">offloading + slicing</td>
        <td style="text-align:center;">2527.651</td>
        <td style="text-align:center;">4.819</td>
    </tr>
</table>

### SDXL


<table>
    <tr>
        <th style="text-align:center;">setting</th>
        <th style="text-align:center;">memory (MB)</th>
        <th style="text-align:center;">inference (secs)</th>
    </tr>
    <tr>
        <td style="text-align:center;">no offloading + no slicing</td>
        <td style="text-align:center;">11238.998</td>
        <td style="text-align:center;">16.826</td>
    </tr>
    <tr>
        <td style="text-align:center;">no offloading + slicing</td>
        <td style="text-align:center;">11245.027</td>
        <td style="text-align:center;">6.817</td>
    </tr>
    <tr>
        <td style="text-align:center;">offloading + no slicing</td>
        <td style="text-align:center;">5588.445</td>
        <td style="text-align:center;">15.736</td>
    </tr>
    <tr>
        <td style="text-align:center;">offloading + slicing</td>
        <td style="text-align:center;">5588.445</td>
        <td style="text-align:center;">15.682</td>
    </tr>
</table>

**Comment by younesbelkada** [comment]
Hi there! I wanted to post some thoughts on this issue to openly brainstorm the results and what we could do next

## On the "slowness" of Bnb-8bit

This is expected, [LLM.int8 (8-bit matmul from bnb)](https://arxiv.org/abs/2208.07339) really shines when you have very large models (>6B) but you will never be faster than fp16 inference because of the overhead the 8-bit matmul decomposition it introduces. Recall that LLM.int8 decomposes the matmul in 2 stages, to perform a part of the matmul in 8-bit and the rest in fp16 with outlier decomposition

![image](https://huggingface.co/blog/assets/96_hf_bitsandbytes_integration/Mixed-int8.gif)

This sadly introduces an overhead, but that overhead becomes less and less considerable for larger models (>6B models). If in the future very large SD models come up, we can consider adding this.  

In addition to that IMO it is a bit of an overkill to use bnb-8bit for diffusers because bnb-8bit aims to mitigating the outlier emergent features for LLMs that I think do not occur for diffusion models (at least at the current model scale). For quantizing in 8bit I think using naive absmax dynamic quantization should do the trick, and this is already supported / covered - https://pytorch.org/blog/accelerating-generative-ai-3/ 

## What about bnb-4bit?

For 4-bit though apprently for  bs=1 [you can get faster inference than fp16]( https://twitter.com/Tim_Dettmers/status/1683118705956491264?s=20) but really depends on the hardware you are using. Apparently you can maximise the speed if you use an RTX4090, if I understood correctly sayak did his experiments on a A100 so if someone can try the same experiment on a RTX4090 using the latest version of bitsandbytes we can complete the experiment with these numbers. 

## Scopes / directions for the future

It is still very interesting IMO to explore quantization on diffusion models to be able to run these models let's say on local computers for anyone (regardless if you have / don't have CPUs, GPUs, etc.). Currently, block-wise 4-bit quantization seems to work well and out of the box on recent LLMs, and if proper kernels are written, it can be easily device-agnostic. I do think this could work out of the box on Linear layers for diffusion models as well. One could use libraries that expose simple quantization primitives (e.g. `QuantLinear`), there are some ongoing discussions on `quanto` library: https://github.com/huggingface/quanto/issues/65 to support 4-bit blockwise quant-linear

**Comment by sayakpaul** [comment]
Thanks so much @younesbelkada! 

Judging by the possibilities and the existing options we have in the library and the ecosystem, I think it's best for us to revisit this in the future. LLM.int8() doesn't directly apply to diffusion models and rightfully so!

**Comment by bigmover** [comment]
> Okay, so I tested the effectiveness of model CPU offload with VAE slicing turned on and off.
> 
> My code
> Bash script to run a bulk of experiments:
> ## Findings
> ### SD
> setting	memory (MB)	inference (secs)
> no offloading + no slicing	3427.855	1.322
> no offloading + slicing	3418.828	1.326
> offloading + no slicing	2526.685	4.078
> offloading + slicing	2527.651	4.819
> ### SDXL
> setting	memory (MB)	inference (secs)
> no offloading + no slicing	11238.998	6.826
> no offloading + slicing	11245.027	6.817
> offloading + no slicing	5588.445	15.736
> offloading + slicing	5588.445	15.682
> I think it might be best for us to not prioritize this issue based on the findings above. Especially if you compare the results from 4bit and 8bit:
> 
> <img alt="image" width="440" src="https://private-user-images.githubusercontent.com/22957388/295535071-cfea716e-f44a-44bb-be52-75c0876e34bc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU2NzA4NDksIm5iZiI6MTcxNTY3MDU0OSwicGF0aCI6Ii8yMjk1NzM4OC8yOTU1MzUwNzEtY2ZlYTcxNmUtZjQ0YS00NGJiLWJlNTItNzVjMDg3NmUzNGJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE0VDA3MDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhNTQ2NTVmNTRjOTgzNzAxZGUwM2IyNjQwOGUxMjg0MGU5MDU4ZjI3ZWIwMGI5ZjhmOGY3NDlhZTI3OTFiNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bryE_LTEr6-HCsstMzHZqjTgFQLU46-gCIcZBCcT0vE">
> @patrickvonplaten WDYT? Cc: @younesbelkada as well.

Hi @sayakpaul  this is an insteresting for qunatination for SD [https://developer.nvidia.com/blog/tensorrt-accelerates-stable-diffusion-nearly-2x-faster-with-8-bit-post-training-quantization/](url) . It will be Nearly 2x Faster. I think It's a good opportunity to support the abilities in diffusers? Any plan?

**Comment by sayakpaul** [comment]
Thanks for bringing this to our attention. The project already makes use of `diffusers` quite heavily and reimplements certain things for better compatibility with TensorRT/ONNX. Plus the following block of code (pulled from the mentioned blog post):

```python
# Load the SDXL-1.0 base model from HuggingFace
import torch
from diffusers import DiffusionPipeline
base = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True
)
base.to("cuda")
 
# Load calibration prompts:
from utils import load_calib_prompts
cali_prompts = load_calib_prompts(batch_size=2,prompts="./calib_prompts.txt")
 
# Create the int8 quantization recipe
from utils import get_percentilequant_config
quant_config = get_percentilequant_config(base.unet, quant_level=3.0, percentile=1.0, alpha=0.8)
 
# Apply the quantization recipe and run calibration  
import ammo.torch.quantization as atq 
quantized_model = atq.quantize(base.unet, quant_config, forward_loop)
 
# Save the quantized model
import ammo.torch.opt as ato
ato.save(quantized_model, 'base.unet.int8.pt')
```

suggests that it should be possible to do int8 quantization with their tooling easily.

**Comment by bigmover** [comment]
> Thanks for bringing this to our attention. The project already makes use of `diffusers` quite heavily and reimplements certain things for better compatibility with TensorRT/ONNX. Plus the following block of code (pulled from the mentioned blog post):
> 
> ```python
> # Load the SDXL-1.0 base model from HuggingFace
> import torch
> from diffusers import DiffusionPipeline
> base = DiffusionPipeline.from_pretrained(
>     "stabilityai/stable-diffusion-xl-base-1.0",
>     torch_dtype=torch.float16,
>     variant="fp16",
>     use_safetensors=True
> )
> base.to("cuda")
>  
> # Load calibration prompts:
> from utils import load_calib_prompts
> cali_prompts = load_calib_prompts(batch_size=2,prompts="./calib_prompts.txt")
>  
> # Create the int8 quantization recipe
> from utils import get_percentilequant_config
> quant_config = get_percentilequant_config(base.unet, quant_level=3.0, percentile=1.0, alpha=0.8)
>  
> # Apply the quantization recipe and run calibration  
> import ammo.torch.quantization as atq 
> quantized_model = atq.quantize(base.unet, quant_config, forward_loop)
>  
> # Save the quantized model
> import ammo.torch.opt as ato
> ato.save(quantized_model, 'base.unet.int8.pt')
> ```
> 
> suggests that it should be possible to do int8 quantization with their tooling easily.

Method by the blog is tied to the depth and NVIDIA ecosystem. For example, they use AMMO to replace torch quant/dequant, and all things will associated with NV engine(restricted the use of LoRa and ControlNet with SD)。
Not so practical！Therefore, I would like to ask Diffusers has any interest for quant of SD？
Thank you for your prompt reply!

**Comment by sayakpaul** [comment]
See: https://github.com/huggingface/diffusers/discussions/7023
